File: ruby/3sum.rb
# frozen_string_literal: true

# https://leetcode.com/problems/3sum
# TODO: fix TLE
# @param {Integer[]} nums
# @return {Integer[][]}
def three_sum(nums)
  return [] if nums.empty?

  triplets = Set.new

  (0...nums.size).each do |i|
    met = Set.new
    (i.next...nums.size).each do |j|
      target = nums[i].+(nums[j]).-@
      triplets.add([nums[i], nums[j], target].sort) if met.include?(target)
      met.add(nums[j])
    end
  end

  triplets.to_a
end

File: javascript/3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
function threeSum(nums) {
    nums.sort((a, b) => a - b);
    const answer = [];
    for (let i = 0; i < nums.length; i++) {
        if (i > 0 && nums[i] === nums[i - 1]) {
            continue;
        }
        twoSum(nums, -nums[i], i + 1, (j, k) => {
            answer.push([nums[i], nums[j], nums[k]]);
        });
    }
    return answer;
}

function twoSum(array, target, from, callback) {
    let l = from;
    let r = array.length - 1;
    while (l < r) {
        while (l > from && array[l] == array[l - 1] && l < r) {
            l++;
        }
        if (l >= r) { break; }
        let sum = array[l] + array[r];
        if (sum < target) {
            l++;
        } else if (sum > target) {
            r--;
        } else {
            callback(l, r);
            l++;
            r--;
        }
    }
}
